general/g.mapsets/list.c

```diff
@@ -35,21 +27,92 @@ void list_accessible_mapsets(const char *fs)
     const char *name;
 
     G_message(_("Accessible mapsets:"));
+
     for (n = 0; (name = G_get_mapset_name(n)); n++) {
         /* match each mapset to its numeric equivalent */
         fprintf(stdout, "%s", name);
         if (G_get_mapset_name(n + 1)) {
-            if (strcmp(fs, "newline") == 0)
-                fprintf(stdout, "\n");
-            else if (strcmp(fs, "space") == 0)
-                fprintf(stdout, " ");
-            else if (strcmp(fs, "comma") == 0)
-                fprintf(stdout, ",");
-            else if (strcmp(fs, "tab") == 0)
-                fprintf(stdout, "\t");
-            else
-                fprintf(stdout, "%s", fs);
+            fprintf(stdout, "%s", fs);
         }
     }
     fprintf(stdout, "\n");
 }
+
+void list_accessible_mapsets_json()
+{
+    int n;
+    const char *name;
+    char *serialized_string = NULL;
+    JSON_Value *root_value = NULL;
+    JSON_Object *root_object = NULL;
+    JSON_Array *mapsets = NULL;
+
+    root_value = json_value_init_object();
+    root_object = json_value_get_object(root_value);
+
+    // Create and add mapsets array to root object
+    json_object_set_value(root_object, "mapsets", json_value_init_array());
+    mapsets = json_object_get_array(root_object, "mapsets");
+
+    // Check that memory was allocated to root json object
+    if (root_value == NULL) {
+        G_fatal_error(_("Failed to initialize JSON. Out of memory?"));
+    }
+
+    // Check that memory was allocated to mapsets array
+    if (mapsets == NULL) {
+        G_fatal_error(_("Failed to initialize JSON array. Out of memory?"));
+    }
+
+    // Add mapsets to mapsets array
+    for (n = 0; (name = G_get_mapset_name(n)); n++) {
+        // Append mapset name to mapsets array
+        json_array_append_string(mapsets, name);
+    }
+
+    // Serialize root object to string and print it to stdout
+    serialized_string = json_serialize_to_string_pretty(root_value);
+    puts(serialized_string);
```

Comment:
Why mix `puts` with the more common (in GRASS code) use of `fprintf`?

Reply:
To me `puts` is cleaner than `fprintf`, but I'm also fine with `fprintf` if that is what the standard is.

Reply:
Well, as far as I'm aware, there is no standard on this regard (rather common practice). The output of `puts` adds a `\n` to the string, which neither `fprintf` nor e.g. `fputs` does not.  In addition, the output stream is by design `stdout`, whereas it is explicitly stated by the others. In my opinion `fprintf(stdout, "%s\n, serialized_string)` may not be so elegant, but have the definitive advantage of being very explicit and the output stream may be easily changed if so wished.
What ever solution you rests with, you should consider make a check if the file operation succeeds.

Reply:
I switched to fprintf and added a check.